### PR TITLE
[USER32] Load keyboard layouts on logon correctly

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
JIRA issue: [CORE-16600](https://jira.reactos.org/browse/CORE-16600)

## Proposed changes

- Add `IntLoadPreloadKeyboardLayouts` helper function to load the layouts on log-on.
- In `UpdatePerUserSystemParameters` function, call `CliImmInitializeHotKeys` and `IntLoadPreloadKeyboardLayouts` functions.

## TODO

- [x] Do tests.

### CODE REVIEW

**Comment by katahiromz** [comment]
Pre-Reboot:
prereboot
I added French and german. And I made French default.

Post-Reboot
postreboot
Correctly restored.

**Comment by HBelusca** [comment]
@katahiromz (Something for a possible separate PR): Could you check why manipulating the layouts would require a reboot?

File: win32ss/user/user32/misc/logon.c [hunk 1]
```diff
@@ -83,6 +84,69 @@ Logon(BOOL IsLogon)
                          sizeof(*LogonRequest));
 }
 
+/* Win: LoadPreloadKeyboardLayouts */
+BOOL IntLoadPreloadKeyboardLayouts(VOID)
+{
+    UINT nNumber, uFlags;
+    DWORD cbValue, dwType;
+    WCHAR szNumber[32], szValue[9];
```
Comment (HBelusca) [inline win32ss/user/user32/misc/logon.c:92]: 9 is certainly `KL_NAMELENGTH` ?

File: win32ss/user/user32/misc/logon.c [hunk 2]
```diff
@@ -83,6 +84,69 @@ Logon(BOOL IsLogon)
...
+                             (LPBYTE)szValue,
+                             &cbValue) != ERROR_SUCCESS)
+        {
+            break;
+        }
+
+        if (dwType != REG_SZ)
+            continue;
+
+        if (nNumber == 1)
```
Comment (HBelusca) [inline win32ss/user/user32/misc/logon.c:122]: Having a comment explaining what this particular case is and why we need different flags here, would be useful.

File: win32ss/user/user32/misc/logon.c [hunk 3]
```diff
@@ -83,6 +84,69 @@ Logon(BOOL IsLogon)
...
+            uFlags = KLF_SUBSTITUTE_OK | KLF_ACTIVATE | KLF_RESET;
+        else
+            uFlags = KLF_SUBSTITUTE_OK | KLF_NOTELLSHELL | KLF_REPLACELANG;
+
+        hKL = LoadKeyboardLayoutW(szValue, uFlags);
+        if (hKL)
+        {
+            ret = TRUE;
```
Comment (HBelusca) [inline win32ss/user/user32/misc/logon.c:130]: It looks like one could get rid of that "ret" variable, and: 1. initialize hKL originally to NULL (where it's defined), and 2. replace all those "ret" checks by suitable checks on the value of hKL; finally, 3. replacing `return ret;` by `return (hKL != NULL);`

File: win32ss/user/user32/misc/logon.c [hunk 4]
```diff
@@ -83,6 +84,69 @@ Logon(BOOL IsLogon)
...
+    if (RegOpenKeyW(HKEY_CURRENT_USER,
+                    L"Keyboard Layout\\Preload",
+                    &hPreloadKey) != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
+
+    for (nNumber = 1; nNumber <= 1000; ++nNumber)
+    {
+        StringCchPrintfW(szNumber, _countof(szNumber), L"%u", nNumber);
```
Comment (HBelusca) [inline win32ss/user/user32/misc/logon.c:106]: Shouldn't using `_ultow` be enough? With your 32-character-sized `szNumber` buffer, there is no way the function could ever do a buffer overrun on it, unless that function itself is buggy (which it shouldn't be).

**Comment by katahiromz** [comment]
> Could you check why manipulating the layouts would require a reboot?

The keyboard layout can be loaded by `LoadKeyboardLayout...` functions. The keyboard layout can be unloaded by `UnloadKeyboardLayout` function. The loaded layouts are stored as the linked list of `gspklBaseLayout` of the kernel subsystem. The user-side application can access the layout list by calling `GetKeyboardLayoutList` function. We can change the input language of the windows by `WM_INPUTLANGCHANGEREQUEST` message broadcasting. The default keyboard layout can be changed `SystemParametersInfo.SPI_SETDEFAULTINPUTLANG`.

I think rebooting is not required to change the layout list.

File: win32ss/user/user32/misc/logon.c (hunk 4 above)
Comment (katahiromz) [reply win32ss/user/user32/misc/logon.c]: Used `_ultow` function.

File: win32ss/user/user32/misc/logon.c (hunk 1 above)
Comment (katahiromz) [reply win32ss/user/user32/misc/logon.c]: Fixed.

File: win32ss/user/user32/misc/logon.c (hunk 2 above)
Comment (katahiromz) [reply win32ss/user/user32/misc/logon.c]: Added a comment.

**Comment by HBelusca** [comment]
> I think rebooting is not required to change the layout list.

One example why requiring reboot to change the layouts is problematic is, for example, during ROS installation. For example, you install ROS in some language (say, japanese), but you don't have the corresponding ideal keyboard layout, and instead you need to use a german keyboard layout to navigate in the installation process (and to use ROS later on). In the installer you can change the keyboard layout, but currently it would ask you to reboot, which is obviously ridiculous at this point, since this would signify you're also aborting the installation.

File: win32ss/user/user32/misc/logon.c (hunk 3 above)
Comment (HBelusca) [reply win32ss/user/user32/misc/logon.c:129]: > I can't agree that. How to check `hKL` without assignment?

> 1. initialize hKL originally to NULL (where it's defined)

i.e. `HKL hKL = NULL, hDefaultKL = NULL;` that is at the top of the function. Then the hKL variable is always correctly defined: to NULL if nothing has been loaded, or to a successfully loaded layout.

Comment (katahiromz) [reply win32ss/user/user32/misc/logon.c:129]: Difficult to see your opinion.

- There can be zero, single or multiple layout(s).
- `LoadKeyboardLayoutW` can be called in multiple times. Not just once.

**Comment by KRosUser** [comment]
Looking at the task bar it seems that the icon is incorrect. Do you confirm ?

**Comment by katahiromz** [comment]
> it seems that the icon is incorrect.

I made French default.

**Comment by katahiromz** [comment]
> I meant task bar not tray icon

Surely. But unrelated to this PR. I will fix that later.

**Comment by KRosUser** [comment]
> > I meant task bar not tray icon
> 
> Surely. But unrelated to this PR. I will fix that later.

Ok, super clear. Thanks!

File: win32ss/user/user32/misc/logon.c (hunk 3 above)
Comment (HBelusca) [reply win32ss/user/user32/misc/logon.c:129]: > LoadKeyboardLayoutW can be called in multiple times. Not just once.

Oh indeed! So you would return TRUE if at least one LoadKeyboardLayoutW call succeed? (even if some others didn't?) If so, then my proposition doesn't make sense, and indeed you should keep the separate `ret` variable.

Comment (katahiromz) [reply win32ss/user/user32/misc/logon.c:129]: The return value is not important because the caller ignores it.

Comment (HBelusca) [reply win32ss/user/user32/misc/logon.c:129]: Then why is there a `ret` variable altogether? The IntLoadPreloadKeyboardLayouts should then just be "void" :smile:

Comment (katahiromz) [reply win32ss/user/user32/misc/logon.c]: Now return void. OK.

**Comment by katahiromz** [comment]
@KRosUser FYI:
https://github.com/reactos/reactos/commit/976d48baeb1a98d620107885022b44b24bad77a1
The icon is fixed.

**Pull request merged**